gateway/oauth_manager_test.go

```diff
@@ -374,6 +376,87 @@ func TestAPIClientAuthorizeAuthCode(t *testing.T) {
 	})
 }
 
+func TestAPIClientAuthorizeRequestWithPKCE(t *testing.T) {
+	ts := StartTest()
+	defer ts.Close()
+
+	spec := LoadAPI(buildTestOAuthSpec(true))[0]
+
+	pID := CreatePolicy(func(p *user.Policy) {
+		p.ID = "TEST-4321"
+		p.AccessRights = map[string]user.AccessDefinition{
+			"test": {
+				APIID: "test",
+			},
+			"abc": {
+				APIID: "abc",
+			},
+		}
+	})
+
+	var redirectURI string
+	// If separator is not set that means multiple redirect uris not supported
+	if config.Global().OauthRedirectUriSeparator == "" {
+		redirectURI = "http://client.oauth.com"
+
+		// If separator config is set that means multiple redirect uris are supported
+	} else {
+		redirectURI = strings.Join([]string{"http://client.oauth.com", "http://client2.oauth.com", "http://client3.oauth.com"}, config.Global().OauthRedirectUriSeparator)
+	}
+	testClient := OAuthClient{
+		ClientID:          authClientID,
+		ClientSecret:      "",
+		ClientRedirectURI: redirectURI,
+		PolicyID:          pID,
+		MetaData:          map[string]interface{}{"foo": "bar", "client": "meta"},
+	}
+	spec.OAuthManager.OsinServer.Storage.SetClient(testClient.ClientID, "org-id-1", &testClient, false)
+
+	t.Run("Client authorize token request with PKCE and no code challenge", func(t *testing.T) {
+		param := make(url.Values)
+		param.Set("response_type", "code")
+		param.Set("client_id", authClientID)
+		param.Set("key_rules", keyRules)
+
+		headers := map[string]string{
+			"Content-Type": "application/x-www-form-urlencoded",
+		}
+
+		_, _ = ts.Run(t, test.TestCase{
+			Path:      "/APIID/tyk/oauth/authorize-client/",
+			AdminAuth: true,
+			Data:      param.Encode(),
+			Headers:   headers,
+			Method:    http.MethodPost,
+			Code:      http.StatusForbidden,
+		})
+	})
```

Comment:
I'm not sure we should enforce this. It's up to the client, not the server.

Reply:
RFC says differently https://tools.ietf.org/html/rfc7636#section-4.4.1

Reply:
So this is an an optional security thing then right? force PKCE.
In which case, should  we have another config option - `enable pkce`?

Reply:
PKCE is there by default so no need to enable. If the code challenge it present then the token call is validated against it for that client